service/s3/internal/customizations/updateEndpoint.go

```diff
@@ -0,0 +1,156 @@
+		}
+	}
+
+	return next.HandleSerialize(ctx, in)
+}
+
+func (u updateEndpointMiddleware) updateEndpointFromConfig(req *http.Request, bucket string) error {
+	// if bucket len is zero, we ignore the following customizations
+	if u.useAccelerate {
+		if u.usePathStyle {
+			// TODO: log that accelerate is not compatible with aws.Config.S3ForcePathStyle, ignoring S3ForcePathStyle.
+		}
+
+		if !hostCompatibleBucketName(req.URL, bucket) {
+			return &smithy.SerializationError{
+				Err: fmt.Errorf("bucket name %s is not compatible with S3 Accelerate", bucket),
+			}
+		}
+
+		parts := strings.Split(req.URL.Host, ".")
+		if len(parts) < 3 {
+			return &smithy.SerializationError{
+				Err: fmt.Errorf("unable to update endpoint host for S3 accelerate, hostname invalid, %s",
+					req.URL.Host),
+			}
+		}
+
+		if parts[0] == "s3" || strings.HasPrefix(parts[0], "s3-") {
+			parts[0] = "s3-accelerate"
+		}
+		for i := 1; i+1 < len(parts); i++ {
+			if parts[i] == u.region {
+				parts = append(parts[:i], parts[i+1:]...)
+				break
+			}
+		}
+		req.URL.Host = strings.Join(parts, ".")
+		moveBucketNameToHost(req.URL, bucket)
+	} else if !u.usePathStyle {
+		if !hostCompatibleBucketName(req.URL, bucket) {
+			// bucket name must be valid to put into the host
+			return &smithy.SerializationError{
+				Err: fmt.Errorf("bucket name %s is not compatible with S3 Accelerate", bucket),
+			}
+		}
+		moveBucketNameToHost(req.URL, bucket)
+	}
+
+	return nil
+}
+
+// updates endpoint to use virtual host styling
+func moveBucketNameToHost(u *url.URL, bucket string) {
+	u.Host = bucket + "." + u.Host
+	removeBucketFromPath(u, bucket)
+}
+
+// remove bucket from url
+func removeBucketFromPath(u *url.URL, bucket string) {
+	u.Path = strings.Replace(u.Path, "/"+bucket, "", -1)
+	if u.Path == "" {
+		u.Path = "/"
+	}
+}
+
+// bucketGetter is an accessor interface to grab the "Bucket" field from
+// an S3 type.
+type bucketGetter interface {
+	GetBucket() string
+}
+
+// hostCompatibleBucketName returns true if the request should
+// put the bucket in the host. This is false if S3ForcePathStyle is
+// explicitly set or if the bucket is not DNS compatible.
+func hostCompatibleBucketName(u *url.URL, bucket string) bool {
+	// Bucket might be DNS compatible but dots in the hostname will fail
+	// certificate validation, so do not use host-style.
+	if u.Scheme == "https" && strings.Contains(bucket, ".") {
+		return false
+	}
+
+	// if the bucket is DNS compatible
+	return dnsCompatibleBucketName(bucket)
+}
+
+var reDomain = regexp.MustCompile(`^[a-z0-9][a-z0-9\.\-]{1,61}[a-z0-9]$`)
+var reIPAddress = regexp.MustCompile(`^(\d+\.){3}\d+$`)
+
+// dnsCompatibleBucketName returns true if the bucket name is DNS compatible.
+// Buckets created outside of the classic region MUST be DNS compatible.
+func dnsCompatibleBucketName(bucket string) bool {
+	return reDomain.MatchString(bucket) &&
+		!reIPAddress.MatchString(bucket) &&
+		!strings.Contains(bucket, "..")
+}
```

Comment:
regex is really slow in Go, this is an area we could consider replacing with a for loop.